### PR TITLE
storage: add an admin command for Import

### PR DESCRIPTION
**Pull request by danhhz**
And use it in the implementation of Restore. This distributes and
parallelizes reading the files in a backup, rekeying them, and sending
out the WriteBatch requests. Import is an admin command, and so runs on
the leaseholder. This means the WriteBatch call will usually be to the
same node (for performance) but it's not required for correctness.

evalImport is mostly code movement of the previous implemention of
sqlccl.Import, with notable exception of the addition of calls to
`validateRangeEmpty` and `beginLimitedRequest`.

Benchmarks don't change locally, but this is only expected to have an
effect on a multi-machine cluster.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13726)

### CODE REVIEW

**Comment by danhhz** [comment]
@bdarnell I made the `storage.SetImportCmd` hook take a `storage.Command` to match the others, but since `Import` is an admin command, it doesn't really need to be. Are you happy with this or would you prefer something else?

And could you please double check my command flags?

**Comment by bdarnell** [comment]
Reviewed 14 of 14 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup.go, line 307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcGFeyYaUgjgF39O8d:-KdcGFeyYaUgjgF39O8e:b-4z9ibz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/sqlccl/backup.go#L307)):*
> ```Go
> 			EndKey: endKey,
> 		},
> 		File:       files,
> ```

You may be able to get DistSender to do more work for you if you make a batch of many ImportRequests, each containing one file.

---

*[pkg/ccl/storageccl/import.go, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcCMbT_Df57bibKvbC:-KdcCMbT_Df57bibKvbD:b-lkanv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L1)):*
> ```Go
> // Copyright 2016 The Cockroach Authors.
> ```

s/6/7/

---

*[pkg/ccl/storageccl/import.go, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcGctbVgrmst3P0YRI:-KdcGctbVgrmst3P0YRJ:bb1tpdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L49)):*
> ```Go
> ) (storage.EvalResult, error) {
> 	args := cArgs.Args.(*roachpb.ImportRequest)
> 	_ = resp.(*roachpb.ImportResponse)
> ```

Just remove this line if you're not using it.

---

*[pkg/ccl/storageccl/import.go, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcDSKgrWfXWxFQPz_U:-KdcDSKgrWfXWxFQPz_V:b-an03ur) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L57)):*
> ```Go
> 	defer cancel()
> 
> 	if err := validateRangeEmpty(ctx, db, args.Span); err != nil {
> ```

This doesn't guarantee that the range will *stay* empty. Do we rely on this guarantee for correctness or is it just protection against shooting yourself in the foot? 

---

*[pkg/ccl/storageccl/import.go, line 71 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcEf4p6jz88ED3BwaR:-KdcEf4qORWf6nxzIW3F:b-5va7ya) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L71)):*
> ```Go
> 	var wg util.WaitGroupWithError
> 	b := struct {
> 		engine.RocksDBBatchBuilder
> ```

Embedding this looks odd; I'd give this field a name (or get rid of the anonymous struct completely? It's not clear what that's doing here)

---

*[pkg/roachpb/api.go, line 907 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcBvzl4sWpFZrKH4L3:-KdcBvzm776JU0teVwUv:bktltl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/roachpb/api.go#L907)):*
> ```Go
> func (*WriteBatchRequest) flags() int               { return isWrite | isRange }
> func (*ExportRequest) flags() int                   { return isRead | isRange }
> func (*ImportRequest) flags() int                   { return isAdmin | isRange }
> ```

I think it should have `isAlone` (all the admin requests should probably have this).

Making this a range request means that the DistSender could split it up if the range boundaries don't match expectations,  and then the command would effectively get duplicated. I think you want to make this a point request instead. (and include the range in the payload for validation)

---

*[pkg/storage/replica_command.go, line 2164 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcB1fTJam3d_18fgR6:-KdcB1fTJam3d_18fgR7:bq8oel5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/storage/replica_command.go#L2164)):*
> ```Go
> // SetImportCmd allows setting the function that will be called as the
> // implementation of the Import command. Only allowed to be called by Init.
> func SetImportCmd(cmd Command) {
> ```

I'd just pass the eval function instead of a full `Command` that won't be used. And then you can change the signature to avoid things like the unused `EvalResult`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726)*

**Comment by tamird** [comment]
Reviewed 12 of 14 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup.go, line 289 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-Kdex0JmNgYeV3dUUhER:-Kdex0JmNgYeV3dUUhES:b-hqmlde) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/sqlccl/backup.go#L289)):*
> ```Go
> 	}
> 
> 	newStartKey, ok := kr.RewriteKey(append([]byte{}, startKey...))
> ```

nit: append([]byte(nil), startKey...)

---

*[pkg/ccl/sqlccl/backup.go, line 293 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-Kdex6B9VwaUw1StGYOb:-Kdex6B9VwaUw1StGYOc:bbl5rtr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/sqlccl/backup.go#L293)):*
> ```Go
> 		return errors.Errorf("could not rewrite key: %s", startKey)
> 	}
> 	newEndKey, ok := kr.RewriteKey(append([]byte{}, endKey...))
> ```

ditto nit: append([]byte(nil), endKey...)

---

*[pkg/ccl/storageccl/import.go, line 39 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdexUwA21HJk4AXkieh:-KdexUwA21HJk4AXkiei:b-aqj1ir) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L39)):*
> ```Go
> 	}
> 	if len(kvs) > 0 {
> 		return errors.New("Import can only be called on empty ranges")
> ```

capitalized error messages don't compose well - uncapitalize please

---

*[pkg/ccl/storageccl/import.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-Kdexh5noF_XoGmyHRcp:-Kdexh5o8njO109JDGYO:b-w3u05r) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L46)):*
> ```Go
> // evalImport bulk loads key/value entries.
> func evalImport(
> 	ctx context.Context, _ engine.ReadWriter, cArgs storage.CommandArgs, resp roachpb.Response,
> ```

s/resp/_/?

---

*[pkg/ccl/storageccl/import.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdexmGR4O5FYQiiqiYR:-KdexmGR4O5FYQiiqiYS:b-3rreuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L52)):*
> ```Go
> 	db := cArgs.Repl.DB()
> 
> 	ctx, span := tracing.ChildSpan(ctx, fmt.Sprintf("import %s-%s", args.DataSpan.Key, args.DataSpan.EndKey))
> ```

nit: `[%s-%s)`, i think (looks like this applies in a few places below, might be better to add a String method to Span and use throughout)

---

*[pkg/ccl/storageccl/import_test.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdeyNnDyHpdrSLtu-Vc:-KdeyNnDyHpdrSLtu-Vd:bkyly8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import_test.go#L36)):*
> ```Go
> 		return err
> 	}
> 	ts := hlc.NewClock(hlc.UnixNano, 0).Now()
> ```

`hlc.NewClock(hlc.UnixNano, time.Nanosecond)` - we try to avoid "zero value" clocks in tests (note that we currently have none)

---

*[pkg/roachpb/api.proto, line 799 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdexFdb1CjIn_Nqcqj1:-KdexFdb1CjIn_Nqcqj2:b-nyvehr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/roachpb/api.proto#L799)):*
> ```protobuf
>   }
>   optional Span header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
>   repeated File file = 2 [(gogoproto.nullable) = false];
> ```

how come `file` and not `files`?

---

*[pkg/roachpb/api.proto, line 801 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdexHbOtvfu6UuAQvL7:-KdexHbOtvfu6UuAQvL8:b-aajgv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/roachpb/api.proto#L801)):*
> ```protobuf
>   repeated File file = 2 [(gogoproto.nullable) = false];
>   optional Span data_span = 3 [(gogoproto.nullable) = false];
>   repeated KeyRewrite key_rewrite = 4 [(gogoproto.nullable) = false];
> ```

ditto, why not plural?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726)*

**Comment by tamird** [comment]
Reviewed 2 of 14 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726)*

**Comment by danhhz** [comment]
Thanks for the reviews! RFAL

---

Review status: 1 of 15 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/ccl/sqlccl/backup.go, line 289 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-Kdex0JmNgYeV3dUUhER:-KdfnkhkTfk9rnWc7OpA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/sqlccl/backup.go#L289)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: append([]byte(nil), startKey...)
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 293 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-Kdex6B9VwaUw1StGYOb:-Kdfnl03xSW40KfrMlXg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/sqlccl/backup.go#L293)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto nit: append([]byte(nil), endKey...)
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcGFeyYaUgjgF39O8d:-Kdg4CzwajdL-8h4SQGk:bhjj9d4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/sqlccl/backup.go#L307)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

You may be able to get DistSender to do more work for you if you make a batch of many ImportRequests, each containing one file.
</blockquote></details>

Later files override earlier ones (incremental backups). I expanded the comments in api.proto to hopefully make this more clear.

---

*[pkg/ccl/storageccl/import.go, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcCMbT_Df57bibKvbC:-Kdfo6UDSY3ATaJmyxDn:b-uzlkxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L1)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/6/7/
</blockquote></details>

I've been doing these based on when I wrote them instead of when they were committed to the public repo. Should I fix the others? (It would be easy, everything in ccl was committed in 2017)

---

*[pkg/ccl/storageccl/import.go, line 39 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdexUwA21HJk4AXkieh:-KdfoR87pztWei9I8rCT:bq0t1rl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L39)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

capitalized error messages don't compose well - uncapitalize please
</blockquote></details>

Even when it's proper noun-ish like this one is? (done though)

---

*[pkg/ccl/storageccl/import.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-Kdexh5noF_XoGmyHRcp:-Kdfob7PJU-iahKdDnKR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L46)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/resp/_/?
</blockquote></details>

Done.

---

*[pkg/ccl/storageccl/import.go, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcGctbVgrmst3P0YRI:-KdfobagVjY7SslU0VDU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L49)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Just remove this line if you're not using it.
</blockquote></details>

Done.

---

*[pkg/ccl/storageccl/import.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdexmGR4O5FYQiiqiYR:-Kdg5Bhj6KAsK9WcJVfY:b-me73z) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L52)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: `[%s-%s)`, i think (looks like this applies in a few places below, might be better to add a String method to Span and use throughout)
</blockquote></details>

I tried to add a String method to Span, but there's already one in the generated proto code. Done

---

*[pkg/ccl/storageccl/import.go, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcDSKgrWfXWxFQPz_U:-Kdfp3FU6ru6X7Bjk9VA:b-o8ett6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L57)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This doesn't guarantee that the range will *stay* empty. Do we rely on this guarantee for correctness or is it just protection against shooting yourself in the foot? 
</blockquote></details>

The latter. Added a comment.

---

*[pkg/ccl/storageccl/import.go, line 71 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcEf4p6jz88ED3BwaR:-KdfqAZbBULNmby9A74S:be9y05g) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L71)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Embedding this looks odd; I'd give this field a name (or get rid of the anonymous struct completely? It's not clear what that's doing here)
</blockquote></details>

Struct is a result of https://reviewable.io/reviews/cockroachdb/cockroach/13586#-Kd76mtdB6yPapU5EABc:-Kd76mtdB6yPapU5EABd:b7lnouc

Un-embedded that field and named the struct so I can use it to reset all three fields at once

---

*[pkg/ccl/storageccl/import_test.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdeyNnDyHpdrSLtu-Vc:-KdfrJ--R_BlogctxtRA:b26rzfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import_test.go#L36)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`hlc.NewClock(hlc.UnixNano, time.Nanosecond)` - we try to avoid "zero value" clocks in tests (note that we currently have none)
</blockquote></details>

Done.

Is this best practice documented somewhere? If it is, I missed it, so maybe we need to make it more prominent somehow

---

*[pkg/roachpb/api.go, line 907 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcBvzl4sWpFZrKH4L3:-KdgCgz9PpQD7lh1ecM3:b-aceea0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/roachpb/api.go#L907)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think it should have `isAlone` (all the admin requests should probably have this).

Making this a range request means that the DistSender could split it up if the range boundaries don't match expectations,  and then the command would effectively get duplicated. I think you want to make this a point request instead. (and include the range in the payload for validation)
</blockquote></details>

Good point. Done. I've already got an issue to test backup and restore in the presence of splits (which would have caught this), so I'll make sure that gets done in a followup PR.

---

*[pkg/roachpb/api.proto, line 799 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdexFdb1CjIn_Nqcqj1:-Kdfy_96rDehDb7D47Z-:bdhmprt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/roachpb/api.proto#L799)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

how come `file` and not `files`?
</blockquote></details>

Whoops. Ingrained from my google days where repeated fields were singular (because of the way protoc generates c++ code)

---

*[pkg/roachpb/api.proto, line 801 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdexHbOtvfu6UuAQvL7:-Kdfyv6Lq2yIEWZ3lF8U:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/roachpb/api.proto#L801)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto, why not plural?
</blockquote></details>

Done

---

*[pkg/storage/replica_command.go, line 2164 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcB1fTJam3d_18fgR6:-KdfyxD4CDpBv_pf2Rc8:bh4jloh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/storage/replica_command.go#L2164)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'd just pass the eval function instead of a full `Command` that won't be used. And then you can change the signature to avoid things like the unused `EvalResult`.
</blockquote></details>

Okay! Thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-:-KdgGpDc1EEnxOsLoXyh:b-21u8wv)*

**Comment by tamird** [comment]
Reviewed 14 of 14 files at r2.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/storageccl/import.go, line 39 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdexUwA21HJk4AXkieh:-KdgKj1mey-ZWdLrAwbN:bqa8fvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L39)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Even when it's proper noun-ish like this one is? (done though)
</blockquote></details>

ah, I didn't realize it was meant to be proper nouh-ish because it's not in a method called "Import"

Can you make this error less "import" specific? I'd also be OK with this helper going away if it's only to be used in Import; having helpers like this lying around can result in nonsensical error messages when they're reused in unexpected places.

---

*[pkg/ccl/storageccl/import.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdexmGR4O5FYQiiqiYR:-KdgLCwz1ezzWg4-HQkL:bzc6kk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L52)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I tried to add a String method to Span, but there's already one in the generated proto code. Done
</blockquote></details>

You can defeat the generated method with `option (gogoproto.goproto_stringer) = false;`

---

*[pkg/ccl/storageccl/import_test.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdeyNnDyHpdrSLtu-Vc:-KdgLb7Fv6Mtg60vDM8j:b-svnbep) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import_test.go#L36)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Done.

Is this best practice documented somewhere? If it is, I missed it, so maybe we need to make it more prominent somehow

</blockquote></details>

It's not - I boiled the ocean at some point when I discovered that various bugs were hiding under the various zero clocks and offsets.

Please do document it if you have a sense of where it should go!

---

*[pkg/roachpb/api.proto, line 889 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdgLvq-mln0lRsNYD8l:-KdgLvq-mln0lRsNYD8m:b4b7g28) ([raw file](https://github.com/cockroachdb/cockroach/blob/8086a2e6e34bd642d99812359475236543760baf/pkg/roachpb/api.proto#L889)):*
> ```protobuf
>   optional ExportRequest export = 32;
>   optional ImportRequest import = 34;
>   optional QueryTxnRequest query_txn = 33;
> ```

can you keep these in tag order?

---

*[pkg/storage/replica_command.go, line 2181 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdgM26qOlq8dHF7M8QA:-KdgM26qOlq8dHF7M8QB:b-hurmo8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8086a2e6e34bd642d99812359475236543760baf/pkg/storage/replica_command.go#L2181)):*
> ```Go
> var exportCmd = makeUnimplementedCommand(roachpb.Export)
> var importCmdFn ImportCmdFunc = func(context.Context, CommandArgs) error {
> 	return errors.Errorf("unimplemented command: %s", roachpb.Import.String())
> ```

.String() is not needed, I believe.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726)*

**Comment by bdarnell** [comment]
:lgtm:

---

Reviewed 14 of 14 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/storageccl/import.go, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcCMbT_Df57bibKvbC:-KdgPy8FMpMH-NANe6UJ:b-ngi4wk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L1)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I've been doing these based on when I wrote them instead of when they were committed to the public repo. Should I fix the others? (It would be easy, everything in ccl was committed in 2017)
</blockquote></details>

Oh, I thought we were done with all the stuff from the other repo and this was new code. I think either date is fine, then.

---

*[pkg/ccl/storageccl/import_test.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdeyNnDyHpdrSLtu-Vc:-KdgR7YF9JLQvGlJ9GXE:b-brk0y8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import_test.go#L36)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

It's not - I boiled the ocean at some point when I discovered that various bugs were hiding under the various zero clocks and offsets.

Please do document it if you have a sense of where it should go!
</blockquote></details>

The comments on `hlc.NewClock` would be the most obvious place. Maybe even make it panic if passed a zero. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-:-KdgS6QFZQmm1ltwlCND:bnfp4nl)*

**Comment by danhhz** [comment]
Review status: 12 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/ccl/storageccl/import.go, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdcCMbT_Df57bibKvbC:-KdgSRGXNWMtCFcCScLu:bwvirnu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L1)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Oh, I thought we were done with all the stuff from the other repo and this was new code. I think either date is fine, then.
</blockquote></details>

This is the second to last piece (the last is the 10 node 2TB acceptance test that mjibson is cherry-picking)

---

*[pkg/ccl/storageccl/import.go, line 39 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdexUwA21HJk4AXkieh:-KdgRoI_3DK3-O2Hk3y_:b-gw3i5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L39)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ah, I didn't realize it was meant to be proper nouh-ish because it's not in a method called "Import"

Can you make this error less "import" specific? I'd also be OK with this helper going away if it's only to be used in Import; having helpers like this lying around can result in nonsensical error messages when they're reused in unexpected places.
</blockquote></details>

I find this cleaner with the start and end key it generates not polluting the evalImport namespace. Made the function name more specific

---

*[pkg/ccl/storageccl/import.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdexmGR4O5FYQiiqiYR:-KdgR_cM5eseGPwAnPFN:bz1ao6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import.go#L52)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

You can defeat the generated method with `option (gogoproto.goproto_stringer) = false;`
</blockquote></details>

good to know. sounds like a different PR though

---

*[pkg/ccl/storageccl/import_test.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdeyNnDyHpdrSLtu-Vc:-KdgTDqf3sLflQRCdNaL:b-z00rvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8ba7d9ca6dca05593ff1800dac955ce900f2ad/pkg/ccl/storageccl/import_test.go#L36)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The comments on `hlc.NewClock` would be the most obvious place. Maybe even make it panic if passed a zero. 
</blockquote></details>

I'll do this. Made an issue so I don't forget https://github.com/cockroachdb/cockroach/issues/13768

---

*[pkg/roachpb/api.proto, line 889 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdgLvq-mln0lRsNYD8l:-KdgRC5xnnxAfpQDhqKC:bdhj5bo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8086a2e6e34bd642d99812359475236543760baf/pkg/roachpb/api.proto#L889)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can you keep these in tag order?
</blockquote></details>

there seems to already be precedent (all over the place in sql and even in this message) for grouping them in some logical non-tag order

---

*[pkg/storage/replica_command.go, line 2181 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726#-KdgM26qOlq8dHF7M8QA:-KdgRXYm6kmmLGnmmLpY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8086a2e6e34bd642d99812359475236543760baf/pkg/storage/replica_command.go#L2181)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

.String() is not needed, I believe.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726)*

**Comment by tamird** [comment]
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726)*

File: pkg/roachpb/api.proto [hunk 1]
```diff
@@ -815,6 +815,31 @@ message ExportResponse {
   repeated File files = 2 [(gogoproto.nullable) = false];
 }
 
+// ImportRequest is the argument to the Import() method, to bulk load key/value
+// entries.
+message ImportRequest {
+  message File {
+    optional ExportStorage dir = 1 [(gogoproto.nullable) = false];
+    optional string path = 3 [(gogoproto.nullable) = false];
```
Comment (dt) [inline pkg/roachpb/api.proto]: what happened to `2` ? should it be reserved?

Comment (danhhz) [inline pkg/roachpb/api.proto]: Done

**Review by dt — APPROVED** [review]
LGTM.

**Comment by tamird** [comment]
Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13726)*

**Comment by danhhz** [comment]
Thanks for the reviews!

**Pull request merged**